## Section 1 — Series and DataFrames

### Task 1.1: Recreate the DataFrame

In [1]:
import pandas as pd

df = pd.DataFrame({
    "name": ["Ali", "Zara", "Jasur"],
    "age": [22, 28, 31],
    "city": ["Tashkent", "Samarkand", "Bukhara"]
})

# Return types
print("df['age'] type:", type(df["age"]))
print("df[['name', 'age']] type:", type(df[["name", "age"]]))

# Print index
print("\nIndex of df:", df.index)

df['age'] type: <class 'pandas.core.series.Series'>
df[['name', 'age']] type: <class 'pandas.core.frame.DataFrame'>

Index of df: RangeIndex(start=0, stop=3, step=1)


**Answers:**
- `df["age"]` returns a **Series**.
- `df[["name", "age"]]` returns a **DataFrame**.
- The index of `df` by default is a `RangeIndex(start=0, stop=3, step=1)`.

### Task 1.2: Add rows and reset index

In [2]:
new_rows = pd.DataFrame({
    "name": ["Malika", "Rustam", "Nodir"],
    "age": [25, 33, 21],
    "city": ["Namangan", "Fergana", "Andijan"]
})

df = pd.concat([df, new_rows], ignore_index=True)
display(df)

,name,age,city
0,Ali,22,Tashkent
1,Zara,28,Samarkand
2,Jasur,31,Bukhara
3,Malika,25,Namangan
4,Rustam,33,Fergana
5,Nodir,21,Andijan


## Section 2 — Loading and Exploring

### Task 2.1: Load a public CSV

In [3]:
# Loading Titanic dataset from a public URL
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
titanic_df = pd.read_csv(url)
display(titanic_df.head(3))

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


### Task 2.2: Inspection Commands

In [4]:
print("--- df.head() ---")
display(titanic_df.head())

print("\n--- df.shape ---")
print(titanic_df.shape)

print("\n--- df.info() ---")
titanic_df.info()

print("\n--- df.describe() ---")
display(titanic_df.describe())

print("\n--- df.isna().sum() ---")
print(titanic_df.isna().sum())

--- df.head() ---


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S



--- df.shape ---
(891, 12)

--- df.info() ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB

--- df.describe() ---


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200



--- df.isna().sum() ---
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


**Inspection Learnings:**
- `df.head()`: I learned what the raw raw values, formatting, and columns of the first 5 records look like (e.g. PassengerId, Survived, Pclass, Name).
- `df.shape`: I learned the dataset has exactly 891 rows and 12 columns.
- `df.info()`: I learned the structural metadata, such as column object types (mostly ints and strings) and memory usage.
- `df.describe()`: I learned the mathematical summary statistics, such as the average fare ($32.2) and the average passenger age (~29.7 years).
- `df.isna().sum()`: I learned exactly how many missing values exist per column, including 177 missing Age values and 687 missing Cabin values.

## Section 3 — Filtering, GroupBy, Merging

### Task 3.1: Selection

In [5]:
# Select one column as a Series
names_series = titanic_df["Name"]

# Select two columns as a DataFrame
age_fare_df = titanic_df[["Age", "Fare"]]

# Use df.loc[] to select a row by its label (Index 0)
row_loc = titanic_df.loc[0]

# Use df.iloc[] to select the first row by position
row_iloc = titanic_df.iloc[0]

display(age_fare_df.head(2))

,Age,Fare
0,22.0,7.2500
1,38.0,71.2833


### Task 3.2: Boolean Filtering

In [6]:
# Filter 1: First-class passengers older than 50
filter_1 = titanic_df[(titanic_df["Pclass"] == 1) & (titanic_df["Age"] > 50)]

# Filter 2: Female passengers OR passengers paying more than $100
filter_2 = titanic_df[(titanic_df["Sex"] == "female") | (titanic_df["Fare"] > 100)]

display(filter_1.head(3))

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
54,55,0,1,"Ostby, Mr. Engelhart Cornelius",male,65.0,0,1,113509,61.9792,B30,C


**Why `&` and Parentheses?**
We must use the bitwise `&` (and `|`) operators in pandas rather than the Python `and` (or `or`) keyword because `and` implicitly attempts to evaluate the *entire* series array's boolean truth value (which is ambiguous), while `&` applies the operation element-wise across the vectorized sequence. Additionally, parentheses `()` are strictly required around conditions because bitwise operators hold higher precedence in Python than comparison operators (like `==` or `>`); without parentheses, it would attempt to evaluate the logic in the wrong order.

### Task 3.3: GroupBy: split-apply-combine

In [7]:
# Pattern 1
mean_age_by_class = titanic_df.groupby("Pclass")["Age"].mean()
print("Mean age by Pclass:")
print(mean_age_by_class)
print("\nReturn type:", type(mean_age_by_class))

# Pattern 2
agg_stats = titanic_df.groupby("Pclass").agg({
    "Age": "mean",
    "PassengerId": "count"
})
print("\nAggregated Stats:")
display(agg_stats)

Mean age by Pclass:
Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64

Return type: <class 'pandas.core.series.Series'>

Aggregated Stats:


,Age,PassengerId
Pclass,,
1,38.233441,216
2,29.877630,184
3,25.140620,491


**What type does the first call return?**
The first call (`df.groupby("some_category")["some_numeric"].mean()`) returns a **pandas Series**, with the grouped category (`Pclass`) acting as the underlying index.

### Task 3.4: Merging

In [8]:
# Build a small lookup DataFrame manually
lookup_df = pd.DataFrame({
    "Pclass": [1, 2, 3],
    "Class_Name": ["First Class", "Second Class", "Third Class"],
    "Dining_Location": ["Upper Deck", "Main Deck", "Lower Deck"]
})

# Merge into main DataFrame
merged_df = pd.merge(titanic_df, lookup_df, on="Pclass", how="left")
display(merged_df[["Name", "Pclass", "Class_Name", "Dining_Location"]].head())

,Name,Pclass,Class_Name,Dining_Location
0,"Braund, Mr. Owen Harris",3,Third Class,Lower Deck
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,First Class,Upper Deck
2,"Heikkinen, Miss. Laina",3,Third Class,Lower Deck
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,First Class,Upper Deck
4,"Allen, Mr. William Henry",3,Third Class,Lower Deck


**What happens to left table rows with no match using "left" join?**
When using a "left" join, any rows in the main (left) DataFrame that fail to find a matching key inside the lookup (right) DataFrame will still be retained in the final output. However, all the resulting columns that were pulled from the right DataFrame for that row will just be populated with `NaN` (null) values since there was no joined data to provide.